projects/plugins/jetpack/modules/masterbar/wp-posts-list/class-posts-list-page-notification.php

```diff
@@ -0,0 +1,122 @@
+ */
+class Posts_List_Page_Notification {
+
+	/**
+	 * Site's Posts page id
+	 *
+	 * @var int|null
+	 */
+	private $posts_page_id;
+
+	/**
+	 * If the Post_list contains the site's Posts Page
+	 *
+	 * @var bool
+	 */
+	private $is_page_in_list = false;
+
+	/**
+	 * Class instance.
+	 *
+	 * @var Posts_List_Page_Notification|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Posts_List_Page_Notification constructor.
+	 *
+	 * @param string $posts_page_id The Posts page configured in WordPress.
+	 */
+	public function __construct( $posts_page_id ) {
+		\add_filter( 'map_meta_cap', array( $this, 'disable_posts_page' ), 10, 4 );
+		\add_filter( 'post_class', array( $this, 'add_posts_page_css_class' ), 10, 3 );
+		\add_action( 'admin_print_footer_scripts-edit.php', array( $this, 'add_notification_icon' ) );
+
+		$this->posts_page_id = '' === $posts_page_id ? null : (int) $posts_page_id;
+	}
+
+	/**
+	 * Creates instance.
+	 *
+	 * @return Posts_List_Page_Notification
+	 */
+	public static function init() {
+		if ( is_null( self::$instance ) ) {
+			self::$instance = new self( \get_option( 'page_for_posts' ) );
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Disable editing and deleting for the page that is configured as a Posts Page.
+	 *
+	 * @param array  $caps Array of capabilities.
+	 * @param string $cap The current capability.
+	 * @param string $user_id The user id.
+	 * @param array  $args Argument array.
+	 * @return array
+	 */
+	public function disable_posts_page( $caps, $cap, $user_id, $args ) {
+		if ( 'edit_post' !== $cap && 'delete_post' !== $cap ) {
+			return $caps;
+		}
+
+		if ( isset( $args[0] ) && $this->posts_page_id === $args[0] ) {
+			$caps[] = 'do_not_allow';
+		}
+
+		return $caps;
+	}
+
+	/**
+	 * Adds a CSS class on the page configured as a Posts Page.
+	 *
+	 * @param array  $classes A list of CSS classes.
+	 * @param string $class A CSS class.
+	 * @param string $post_id The current post id.
+	 * @return array
+	 */
+	public function add_posts_page_css_class( $classes, $class, $post_id ) {
+		if ( $this->posts_page_id !== $post_id ) {
+			return $classes;
+		}
+
+		$this->is_page_in_list = true;
+
+		$classes[] = 'posts-page';
+
+		return $classes;
+	}
+
+	/**
+	 * Add a info icon on the Posts Page letting the user know why they cannot delete and remove the page.
+	 */
+	public function add_notification_icon() {
+		// No need to add the JS since the site is not configured with a Posts Page or the current listview doesn't contain the page.
+		if ( null === $this->posts_page_id || ! $this->is_page_in_list ) {
+			return;
+		}
+
+		$text_notice = __( 'The content of your latest posts page is automatically generated and cannot be edited', 'jetpack' );
+		?>
+		<script>
+			document.querySelector(".posts-page .check-column").innerHTML = '<span title="<?php echo esc_html( $text_notice ); ?>" style="margin-left: 8px" class="dashicons dashicons-info-outline"></span>';
+		</script>
+		<?php
+	}
+}
```

Comment:
Can we use this 
https://developer.wordpress.org/reference/hooks/manage_post_type_posts_columns/

to hook into the checkbox rather than using javascript?

Reply:
I don't think it can because, from what I understood, the `manage_{$post_type}_posts_columns` is used for column headings.

Reply:
You are correct, this filter only alters column headings. What if we didn't alter the checkbox (which doesn't appear for users not having the caps to edit/delete) and instead hook into the title for in that specific screen. 

Like this
![](https://cln.sh/J7bTr3+)

Reply:
Here is the filter docs
https://developer.wordpress.org/reference/hooks/the_title/

Reply:
Unfortunately WP_Posts_List_table uses `draft_or_post_title()` function [which escapes the HTML](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/template.php#L1981) :(. That was my first idea as well while working on it, since in Calypso the icon is placed right after the title.

Reply:
I see! Ok, last sprint with `display_post_states` filter https://developer.wordpress.org/reference/hooks/display_post_states/

![](https://cln.sh/GvwIKT+)

Reply:
This is how these are created
![](https://cln.sh/stUgUA+)

Reply:
Yeah, I've tried with that filter as well but the states are comma separated (we can't remove the comma) and it looked a bit off (tried at the end and in the beginning of the array). We could've remove Core's `Posts Page` state but I thought that might confuse users familiar with WP's UI. 

![Screenshot 2021-05-21 at 10 40 15](https://user-images.githubusercontent.com/995955/119100759-f78f1000-ba20-11eb-8c90-7ee173413248.png)
![Screenshot 2021-05-21 at 10 42 48](https://user-images.githubusercontent.com/995955/119101161-5ce30100-ba21-11eb-9cd8-6acf4404042e.png)

Reply:
Both in the front or in the end of these `post_states` seem a good tradeoff for not using a workaround here for me.

I would be interested though in what @olaolusoga and the rest of @Automattic/ajax have in mind about this!

Reply:
Just for reference, this is how it looks right now (with the hackish method - adding the icon in the page via JS):

![Screenshot 2021-05-21 at 11 57 50](https://user-images.githubusercontent.com/995955/119111671-f31c2480-ba2b-11eb-9f0d-a2c22805d9de.png)

VS adding the icon as a post_state: 

![Screenshot 2021-05-21 at 10 40 15](https://user-images.githubusercontent.com/995955/119100759-f78f1000-ba20-11eb-8c90-7ee173413248.png)
![Screenshot 2021-05-21 at 10 42 48](https://user-images.githubusercontent.com/995955/119101161-5ce30100-ba21-11eb-9cd8-6acf4404042e.png)

We can also remove Core's `Posts Page` state (but I think we shouldn't since users familiar with WP might be a bit confused).

Reply:
Would be the post lock feature useful here? Seems that post slocked with `wp_set_post_lock` cannot be edited and display a locked icon in the table list instead of the checbox:

https://developer.wordpress.org/reference/functions/wp_set_post_lock/
https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/class-wp-posts-list-table.php#L999

If that's too complex or not doable, I prefer the current approach of replacing the checkbox column since it's more user-friendly IMO.

Reply:
I've looked into wp_set_post_lock while researching ways to restrict the access, but unfortunately the message cannot be customised. 

The lock is intended to be used with a (current) user (and I think it auto-expires, at some point). In theory, we could look into a 'System' pseudo-user, but there might some unintended behaviours. Plus, we'll be displaying a message like `System is currently editing`[[0]](https://core.trac.wordpress.org/browser/trunk/src/wp-admin/includes/class-wp-posts-list-table.php#L1074) which might confuse users.

Reply:
Fair enough, let's keep the current JS approach then 👍 (which is BTW super clever, good job!).

Reply:
@BogdanUngureanu thanks for responding to my nudges and @mmtr for chiming in! Let's move on with the JS approach, we can iterate in the future in case it's needed.